vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java

```diff
@@ -20,6 +39,207 @@
+                    .filter(widget -> Objects.equals(detachedElement,
+                            widget.getElement()))
+                    .findAny().ifPresent(detachedWidget -> {
+                        // The child was removed from the dashboard
+
+                        // Remove the registration for the child detach listener
+                        childDetachListenerMap.get(detachedWidget.getElement())
+                                .remove();
+                        childDetachListenerMap
+                                .remove(detachedWidget.getElement());
+
+                        doRemoveWidget(detachedWidget);
+                        updateClient();
+                    });
+        }
+    };
+
+    @Override
+    protected void onAttach(AttachEvent attachEvent) {
+        super.onAttach(attachEvent);
+        updateClient();
+    }
+
+    private void updateClient() {
+        if (pendingUpdate) {
+            return;
+        }
+        pendingUpdate = true;
+        getElement().getNode()
+                .runWhenAttached(ui -> ui.beforeClientResponse(this, ctx -> {
+                    doUpdateClient();
+                    pendingUpdate = false;
+                }));
+    }
+
+    private void doUpdateClient() {
+        getElement().getChildren().forEach(child -> {
+            if (!childDetachListenerMap.containsKey(child)) {
+                childDetachListenerMap.put(child,
+                        child.addDetachListener(childDetachListener));
+            }
+        });
+        List<Integer> widgetNodeIds = getWidgets().stream()
+                .map(this::getWidgetNodeId).toList();
+        getElement().setPropertyList("virtualChildNodeIds", widgetNodeIds);
+        getElement().setPropertyJson("items", createItemsJsonArray());
+    }
+
+    private void attachRenderer() {
+        getElement().executeJs(
+                "Vaadin.FlowComponentHost.patchVirtualContainer(this);");
+        String appId = UI.getCurrent().getInternals().getAppId();
+        getElement().executeJs(
+                "this.renderer = (root, _, model) => Vaadin.FlowComponentHost.setChildNodes($0, [model.item.nodeid], root);",
+                appId);
+    }
+
+    private JsonArray createItemsJsonArray() {
+        JsonArray jsonItems = Json.createArray();
+        for (DashboardWidget widget : widgets) {
+            JsonObject jsonItem = Json.createObject();
+            jsonItem.put("nodeid", getWidgetNodeId(widget));
+            jsonItems.set(jsonItems.length(), jsonItem);
+        }
+        return jsonItems;
+    }
+
+    private int getWidgetNodeId(DashboardWidget widget) {
+        return widget.getElement().getNode().getId();
+    }
+
+    private void doRemoveAllWidgets() {
+        widgets.clear();
+        getElement().removeAllChildren();
+    }
+
+    private void doRemoveWidget(DashboardWidget widget) {
+        widgets.remove(widget);
+        getElement().removeChild(widget.getElement());
```

Comment:
Could `removeVirtualChild` be used instead?

Reply:
Done.